scala/private/phases/api.bzl

```diff
@@ -0,0 +1,51 @@
+load(
+    "@io_bazel_rules_scala//scala:providers.bzl",
+    _ScalaRulePhase = "ScalaRulePhase",
+)
+
+def _adjust_phases(phases, adjustments):
+    if len(adjustments) == 0:
+        return phases
+    phases = phases[:]
+    for (relation, peer_name, name, function) in adjustments:
+        for idx, (needle, _) in enumerate(phases):
+            if relation in ["^", "first"]:
+                phases.insert(0, (name, function))
+            elif relation in ["$", "last"]:
+                phases.append((name, function))
+            elif needle == peer_name:
+                if relation in ["-", "before"]:
+                    phases.insert(idx, (name, function))
+                elif relation in ["+", "after"]:
+                    phases.insert(idx + 1, (name, function))
+                elif relation in ["=", "replace"]:
+                    phases[idx] = (name, function)
+    return phases
+
+def run_phases(ctx, builtin_customizable_phases, fixed_phase):
+    phase_providers = [
+        p[_ScalaRulePhase]
+        for p in ctx.attr._phase_providers
```

Comment:
many places call provider as p, should it be provider to disabmiguate from phase?

Reply:
I suppose you meant the other way around "should it be phase to disabmiguate from provider?". I also worry I might forget what all this `p`s mean later, I decided to use more semantic name in this file.